presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java

```diff
@@ -702,6 +713,51 @@ private void generateHasFilterFunctionMethod(ClassDefinition classDefinition, Fi
                 .ret(boolean.class);
     }
 
+    private void generateApplyFilterFunction(ClassDefinition classDefinition, FieldDefinition joinFilterFunctionField)
+    {
+        Parameter leftBlockIndex = arg("leftBlockIndex", int.class);
+        Parameter leftPosition = arg("leftPosition", int.class);
+        Parameter rightPosition = arg("rightPosition", int.class);
+        Parameter allRightBlocks = arg("allRightBlocks", Block[].class);
+
+        MethodDefinition applyFilterFunctionMethod = classDefinition.declareMethod(
+                a(PUBLIC),
+                "applyFilterFunction",
+                type(boolean.class),
+                leftBlockIndex,
+                leftPosition,
+                rightPosition,
+                allRightBlocks);
+
+        Variable thisVariable = applyFilterFunctionMethod.getThis();
+        applyFilterFunctionMethod.getBody()
+                .append(thisVariable
+                        .getField(joinFilterFunctionField)
+                        .invoke("filter", boolean.class, leftPosition, thisVariable.invoke("getLeftBlocks", Block[].class, leftBlockIndex), rightPosition, allRightBlocks))
+                .retBoolean();
+    }
+
+    private void generateGetLeftBlocksMethod(ClassDefinition classDefinition, FieldDefinition channelArraysField, FieldDefinition emptyBlockArrayField)
+    {
+        Parameter leftBlockIndex = arg("leftBlockIndex", int.class);
+        MethodDefinition getLeftBlocksMethod = classDefinition.declareMethod(
+                a(PUBLIC),
+                "getLeftBlocks",
+                type(Block[].class),
+                leftBlockIndex);
+
+        Variable thisVariable = getLeftBlocksMethod.getThis();
```

Comment:
I think you can check beforehand if joinfilterfunction has some field references. In such case you could remote this branch